lib/Sema/AvailableFactsAnalysis.cpp

```diff
@@ -0,0 +1,458 @@
+     delete Blocks.back();
+     Blocks.pop_back();
+   }
+}
+
+AvailableFactsAnalysis::ElevatedCFGBlock* AvailableFactsAnalysis::GetBlock(std::vector<ElevatedCFGBlock *>& Blocks, CFGBlock *I) {
+  if (BlockIDs[I->getBlockID()] != -1)
+    return Blocks[BlockIDs[I->getBlockID()]];
+  return UnreachableBlock;
+}
+
+void AvailableFactsAnalysis::Reset() {
+  CurrentIndex = 0;
+}
+
+void AvailableFactsAnalysis::Next() {
+  CurrentIndex++;
+}
+
+// This function fills `ComparisonFacts` with pairs (Expr1, Expr2) where
+// Expr1 <= Expr2.
+// These comparisons correspond to the current block.
+void AvailableFactsAnalysis::GetFacts(std::pair<ComparisonSet, ComparisonSet> &CFacts) {
+  CFacts = Facts[CurrentIndex];
+}
+
+// Given two sets S1 and S2, the return value is S1 - S2.
+ComparisonSet AvailableFactsAnalysis::Difference(ComparisonSet &S1, ComparisonSet &S2) {
+ if (S2.size() == 0)
+    return S1;
+  ComparisonSet Result;
+  for (auto E1 : S1)
+    if (S2.find(E1) == S2.end())
+      Result.insert(E1);
+  return Result;
+}
+
+// Given two sets S1 and S2, the return value is the union of these sets.
+ComparisonSet AvailableFactsAnalysis::Union(ComparisonSet &S1, ComparisonSet &S2) {
+  if (S1.size() == 0)
+    return S2;
+  if (S2.size() == 0)
+    return S1;
+  ComparisonSet Result(S1);
+  Result.insert(S2.begin(), S2.end());
+  return Result;
+}
+
+// Given two sets S1 and S2, this function returns true if the two sets
+// are not equal.
+// Equality means that their sizes are the same, and every member of S1 is found in
+// S2 and vice versa.
+bool AvailableFactsAnalysis::Differ(ComparisonSet &S1, ComparisonSet &S2) {
+  if (S1.size() != S2.size())
+    return true;
+  for (auto E : S1)
+    if (S2.find(E) == S2.end())
+      return true;
+  return false;
+}
+
+// Given two sets S1 and S2, the return value is the intersection of these sets.
+ComparisonSet AvailableFactsAnalysis::Intersect(ComparisonSet &S1, ComparisonSet &S2) {
+  if (S1.size() == 0)
+    return S1;
+  if (S2.size() == 0)
+    return S2;
+  ComparisonSet Result;
+  for (auto E1 : S1)
+    if (S2.find(E1) != S2.end())
+      Result.insert(E1);
+  return Result;
+}
+
+// This function returns true if variable `V` is used in the comparison `I`.
+bool AvailableFactsAnalysis::ContainsVariable(Comparison& I, const VarDecl *V) {
+  std::set<const Expr *> Exprs;
+  CollectExpressions(I.first, Exprs);
+  CollectExpressions(I.second, Exprs);
+  for (auto InnerExpr : Exprs)
+    if (const DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(InnerExpr))
+      if (const VarDecl *VD = dyn_cast<VarDecl>(DRE->getDecl()))
+        if (VD == V)
+          return true;
+  return false;
+}
+
+// This function returns true only if an expression in Comparison `I` contains
+// a pointer deref. A pointer deref can appear as one of the following form:
+// 1. With Deref operator `*`
+// 2. `->` for accessing a member
+// 3. Array subscript `[]`
+bool AvailableFactsAnalysis::ContainsPointerDeref(Comparison& I) {
```

Comment:
This isn't the right test for whether an expression is or contains a pointer dereference.  The expression should be a cast expression whose operator is an LValueToRValue cast and whose child expression produces an lvalue from a ponter.  For the child expression, the function f(e) must be true, where f(e) is defined as:

1. If e is one of the expressions you list (*, -> or subscript expr), return true.
2.  If e has the form (e1), return f(e1).
3.  If e is an  lvalue cast or no-op cast of e1, return f(e1).  An lvalue is a cast expression whose operator is LValueBitCast.  The no-op cast operator is NoOp (see clang\include\clang\AST\OperationKinds.def).
4.  Otherwise, return false.